### PR TITLE
Misplaced queue.finish() ???

### PR DESCRIPTION
**Pull request by jean-m-cyr**
Shouldn't this call be inside the work loop?
If not, what synchronizes the work loop with the GPU?

### CODE REVIEW

**Comment by jean-m-cyr** [comment]
I believe this is why you get impossibly high hash rate counts when you increase --cl-global-work too much. You get multiple passes around the work loop, for each GPU kernel executions.

**Comment by chfast** [comment]
I don't think this is the case.

It was added just because if CPU want to break the loop, GPU might have still accessing some buffers.
Every kernel execution works as finish() as well.

But the loop is quite complicated to reason about, so maybe something has changes over time.

**Comment by jean-m-cyr** [comment]
A simple test will demonstrate the failure. On an AMD GPU run with a very large cl-global-work number, say 100000 time your --cl-local-work size. Ethminer will starts showing crazy hash rates. Whith the finish() call inside the loop the behavior returns to normal. What I don't understand is what is synchronizing the loop with low --cl-global-work settings? I don't see any synchronous calls unless there is a job change, or a result is found!

**Comment by jean-m-cyr** [comment]
Found it. There's a synchronous read, done with each pass of the loop, that would slow it down enough to allow small GPU work requests to complete on time.

**Comment by akostadinov** [comment]
For me these settings with `--cl-global-work 1048576 --cl-local-work 256` reports above 130MH/s on Vega FE while actual pool hash rate appears ~13MH/s. `--cl-global-work 524288 --cl-local-work 256` shows 66MH/s while actual differs. So you are right that something is wrong but I've no idea what. Also I don't know how situation changed with #897. Just FYI.

--  **0.15.0.dev0+git.26a51df**

**Comment by smurfy** [comment]
I think the way this PR changes it is wrong.
I added it there where it currently is, because ive changed the ``while(true)`` to ``while(shouldStop)``.
Calling ``queue.finish()``  and similiar in CUDA was required, that the GPU actually stops mining. on ``farm.stop()``.

See my commit:
https://github.com/ethereum-mining/ethminer/commit/5de96142ea312de54bef4f6961f10971a0e0d344

**Comment by smurfy** [comment]
I can't say anything regarding hashrate or if queue.finish should be always be called or not on its new position.

I just can mention, that i moved it there for a reason :) which was not to fix any hashrate related problems.

**Comment by jean-m-cyr** [comment]
Ah, ok. It should also be called once when the loop exits.

**Comment by jean-m-cyr** [comment]
Yes, it should also be called upon exit.

Take a look at PR #897. Unrelated but there is a complaint there about impossible hash rates.

**Comment by chfast** [comment]
Added to v0.14.0rc3: https://github.com/ethereum-mining/ethminer/releases/tag/v0.14.0rc3.

File: libethash-cl/CLMiner.cpp [hunk 1]
```diff
@@ -377,10 +377,11 @@ void CLMiner::workLoop()
 
 			// Report hash count
 			addHashCount(m_globalWorkSize);
-		}
 
-		// Make sure the last buffer write has finished --
-		// it reads local variable.
+			// Make sure the last buffer write has finished --
+			// it reads local variable.
+			m_queue.finish();
```
Comment (akostadinov) [inline libethash-cl/CLMiner.cpp]: Sorry, a little late comment, but shouldn't this finish call be placed before `addHashCount`? Sounds like we want to add hash count only after queue is processed. I didn't investigate the code, just from the look of it.

**Comment by SukharevPavel** [comment]
>  If not, what synchronizes the work loop with the GPU?

`m_queue.enqueueReadBuffer(m_searchBuffer, CL_TRUE, 0, sizeof(results), &results);` with CL_TRUE parameter served for queue synchronization before this commit. Now we can change it to CL_FALSE.

Also, FYI, wrong hash results are received if time for enqueueNDRangeKernel execution is more than 1 second. That was the case for wrong hashrate logging, and error is somewhere in logger itself, not in CLMiner.cpp code.

**Comment by jean-m-cyr** [comment]
I don't think so... The read needs to remain synchronous since the results of the read are used in the next statement. Also a synchronous read here does nothing to sync the GPU with the CPU.

**Comment by jean-m-cyr** [comment]
No, a synchronous read only blocks till the read operation is complete (the data transferred to local memory).

**Comment by SukharevPavel** [comment]
https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clCreateCommandQueue.html

> If the CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE property of a command-queue is not set, the commands enqueued to a command-queue execute in order. For example, if an application calls clEnqueueNDRangeKernel to execute kernel A followed by a clEnqueueNDRangeKernel to execute kernel B, the application can assume that kernel A finishes first and then kernel B is executed. If the memory objects output by kernel A are inputs to kernel B then kernel B will see the correct data in memory objects produced by execution of kernel A. If the CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE property of a commandqueue is set, then there is no guarantee that kernel A will finish before kernel B starts execution. 

In CLMiner.cpp, we don't set CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE flag.

**Comment by jean-m-cyr** [comment]
That applies to consecutive clEnqueueNDRangeKernel calls. The code in CLMiner clearly attempts to overlap GPU and CPU processing, which would be futile if clEnqueueNDRangeKernel did not return immediately (the operation considered complete) upon the launch of the kernel.

**Comment by jean-m-cyr** [comment]
If what you say is true then the sync read would be sufficient to sync the loop. I have always considered an async operation to be dequeued once initiated, I may be wrong..

**Comment by SukharevPavel** [comment]
> That applies to consecutive clEnqueueNDRangeKernel calls. 

In fact, I've just made simple time logging (for code without clFinish between enqueueNDRangeKernel and enqueueReadBuffer):

```
milliseconds lastTime;
	
long printTime(){
	milliseconds time = duration_cast< milliseconds >(
		system_clock::now().time_since_epoch());
	long diff = time.count() - lastTime.count();
	lastTime = time;
	return diff;
};
```

and here is sample results:
```
 cl  17:40:47|cl-0    |  0  right before enqueueNDRangeKernel
 cl  17:40:47|cl-0    |  0  right after enqueueNDRangeKernel
 cl  17:40:47|cl-0    |  0  right before enqueueReadBuffer:
 cl  17:40:47|cl-0    |  92  right after enqueueReadBuffer:
 cl  17:40:47|cl-0    |  0  right before enqueueNDRangeKernel
 cl  17:40:47|cl-0    |  0  right after enqueueNDRangeKernel
 cl  17:40:47|cl-0    |  0  right before enqueueReadBuffer:
 cl  17:40:47|cl-0    |  90  right after enqueueReadBuffer:
```

Also, we can measure enqueueNDRangeKernel and enqueueReadBuffer execution time via OpenCL events. For me, enqueueReadBuffer is exactly about 90 ms, and enqueueReadBuffer is relatively small (I don't remember exact number, but it like some microseconds).

Anyway, all my experiments are conclude that OpenCL queue enqueues both enqueueNDRangeKernel and Read/Write operation in the same way. So, I think, adding clFinish between enqueueNDRangekernel and blocking enqueueReadBuffer is just obsolete.

**Comment by jean-m-cyr** [comment]
Interesting! Based on that, if you double the size of the global work multiplier you should also see a doubling of the sync read delay. Right?

**Comment by jean-m-cyr** [comment]
I can't do any ETH testing right now, All my miners are setup for a different coin.

**Comment by SukharevPavel** [comment]
Yep, the results for  --cl-global-work 16384
```
 cl  18:01:01|cl-0    |  0  right before enqueueNDRangeKernel
 cl  18:01:01|cl-0    |  0  right after enqueueNDRangeKernel
 cl  18:01:01|cl-0    |  0  right before enqueueReadBuffer:
 cl  18:01:01|cl-0    |  188  right after enqueueReadBuffer:
 cl  18:01:01|cl-0    |  0  right before enqueueNDRangeKernel
 cl  18:01:01|cl-0    |  0  right after enqueueNDRangeKernel
 cl  18:01:01|cl-0    |  0  right before enqueueReadBuffer:
 cl  18:01:02|cl-0    |  176  right after enqueueReadBuffer:
```
Also, this tests both were made for `./ethminer -G -M`

**Comment by jean-m-cyr** [comment]
Ok, I'm convinced. The finish() call is redundant....

**Pull request merged**